lib/std/priority_dequeue.zig

```diff
@@ -0,0 +1,879 @@
+const std = @import("std");
+const Allocator = std.mem.Allocator;
+const sort = std.sort;
+const assert = std.debug.assert;
+const warn = std.debug.warn;
+const testing = std.testing;
+const expect = testing.expect;
+const expectEqual = testing.expectEqual;
+const expectError = testing.expectError;
+
+/// Priority Dequeue for storing generic data. Initialize with `init`.
+pub fn PriorityDequeue(comptime T: type) type {
+    return struct {
+        const Self = @This();
+
+        items: []T,
+        len: usize,
+        allocator: *Allocator,
+        lessThanFn: fn (a: T, b: T) bool,
+
+        /// Initialize and return a new dequeue. Provide `lessThanFn`
+        /// that returns `true` when its first argument should
+        /// get min-popped before its second argument. For example,
+        /// to make `popMin` return the minimum value, provide
+        ///
+        /// `fn lessThanFn(a: T, b: T) bool { return a < b; }`
+        pub fn init(allocator: *Allocator, lessThanFn: fn (T, T) bool) Self {
```

Comment:
Design question, what about making the comparator a `fn (T, T) math.Order` ?

Reply:
Thanks for the great review!

I agree that this would be a better API; I was not aware of `math.Order`. If I change this here I think I should change `std.PriorityQueue` as well. This would be a breaking change, but presumably this is common for Zig right now. I will make this change unless other reviewers disagree.

There might be an argument for making similar changes to `std.sort` for consistency, as some functions in there use `bool`, and some use `math.Order`. For example:

```
pub fn binarySearch(
    comptime T: type,
    key: T,
    items: []const T,
    context: anytype,
    comptime compareFn: fn (context: @TypeOf(context), lhs: T, rhs: T) math.Order,
) 
```

```
pub fn sort(
    comptime T: type,
    items: []T,
    context: anytype,
    comptime lessThan: fn (context: @TypeOf(context), lhs: T, rhs: T) bool,
)
```

However that seems like a bigger change and is probably out of scope here.